File: matlab/stereoDP_LD.m
% Stereo Matching using Dynamic Programming (with Left-Disparity Axes DSI)
% Computes a disparity map from a rectified stereo pair using Dynamic Programming

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
Pocc = 5; %occlusion penalty

% Define data cost computation
dataCostComputation = @(left,right) abs(left-right); %absolute differences
%dataCostComputation = @(left,right) (left-right).^2; %square differences

% Define smoothness cost computation
smoothnessCostComputation = @(differences) Pocc*abs(differences);
%smoothnessCostComputation = @(differences) Pocc*min(abs(differences),2); %alternative

% Start timer
timerVal = tic();

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Convert to int32
leftImg = int32(leftImg);
rightImg = int32(rightImg);

% Compute pixel-based matching cost (data cost)
dataCost = zeros(rows,cols,dispLevels,'int32');
for d = 0:dispLevels-1
    %rightImgShifted = shiftArray(rightImg,[0,d]);
    rightImgShifted = circshift(rightImg,d,2); %less accurate, better performances
    dataCost(:,:,d+1) = dataCostComputation(leftImg,rightImgShifted);
end

% Compute smoothness cost
d = 0:dispLevels-1;
smoothnessCost = smoothnessCostComputation(d-d.');
smoothnessCost3d = zeros(1,dispLevels,dispLevels,'int32');
smoothnessCost3d(1,:,:) = smoothnessCost;

D = zeros(rows,cols,dispLevels,'int32'); %minimum costs
T = zeros(rows,cols,dispLevels,'int32'); %transitions
dispMap = zeros(rows,cols);

% Compute DP table (forward pass)
for x = 2:cols
    cost = dataCost(:,x-1,:)+D(:,x-1,:);
    [cost,ind] = min(cost+smoothnessCost3d,[],3);
    D(:,x,:) = cost;
    T(:,x,:) = ind;
end

% Compute disparity map (backtracking)
[~,d] = min(D(:,cols,:),[],3);
for x = cols:-1:1
    dispMap(:,x) = d-1;
    linInd = sub2ind(size(T),(1:rows).',x*ones(rows,1),d);
    d = T(linInd);
end

% Normalize the disparity map for display
scaleFactor = 256/dispLevels;
dispImg = uint8(dispMap*scaleFactor);

% Show disparity map
figure; imshow(dispImg)

% Save disparity map
imwrite(dispImg,'disparityDP_LD.png')

% Stop timer and display running time
elapsedTime = toc(timerVal);
fprintf('Running time: %.2f seconds\n',elapsedTime)

File: python/stereoDP_LD.py
# Stereo Matching using Dynamic Programming (with Left-Disparity Axes DSI)
# Computes a disparity map from a rectified stereo pair using Dynamic Programming

import time
import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt
from genericFunctions import *

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
Pocc = 5 #occlusion penalty

# Define data cost computation
dataCostComputation = lambda left,right: np.absolute(left-right) #absolute differences
#dataCostComputation = lambda left,right: (left-right)**2 #square differences

# Define smoothness cost computation
smoothnessCostComputation = lambda differences: Pocc*np.absolute(differences)
#smoothnessCostComputation = lambda differences: Pocc*np.minimum(np.absolute(differences),2) #alternative

# Start timer
timerVal = time.time()

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Convert to int32
leftImg = leftImg.astype(np.int32)
rightImg = rightImg.astype(np.int32)

# Compute pixel-based matching cost (data cost)
dataCost = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    rightImgShifted = shiftRight(rightImg,d,0)
    dataCost[:,:,d] = dataCostComputation(leftImg,rightImgShifted)

# Compute smoothness cost
d = np.arange(dispLevels)
smoothnessCost = smoothnessCostComputation(d-d[np.newaxis,:].T)
smoothnessCost3d = smoothnessCost[np.newaxis,:,:].astype(np.int32)

D = np.zeros((rows,cols,dispLevels),dtype=np.int32) #minimum costs
T = np.zeros((rows,cols,dispLevels),dtype=np.int32) #transitions
dispMap = np.zeros((rows,cols))

# Compute DP table (forward pass)
for x in range(1,cols):
    cost = dataCost[:,x-1,:]+D[:,x-1,:]
    cost = cost[:,np.newaxis,:]+smoothnessCost3d
    D[:,x,:] = np.amin(cost,axis=2)
    T[:,x,:] = np.argmin(cost,axis=2)

# Compute disparity map (backtracking)
d = np.argmin(D[:,-1,:],axis=1)
for x in range(cols-1,-1,-1):
    dispMap[:,x] = d
    d = T[np.arange(rows),x,d]

# Normalize the disparity map for display
scaleFactor = 256/dispLevels
dispImg = (dispMap*scaleFactor).astype(np.uint8)

# Show disparity map
plt.imshow(dispImg,cmap="gray")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparityDP_LD.png",dispImg)

# Stop timer and display running time
elapsedTime = time.time()-timerVal
print("Running time: {:.2f} seconds".format(elapsedTime))

plt.show()
